### PR TITLE
Core/Mechanic: Fix damage before model contact

### LINKED ISSUE #5794: damage before model contact ?
ive watched out this some times and the thing is:

on retail the melee attacks damage the npc/player at model hit so  if the weapon has contact to the body of the target. but in emus the damage already arrived during the player/npc is swinging his weapon....

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65093-damage-before-model-contact?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Aokromes**
**Changes proposed**:

on retail the melee attacks damage the npc/player at model hit so if the weapon has contact to the body of the target. but in emus the damage already arrived during the player/npc is swinging his weapon....

**Target branch(es)**: 335

**Issues addressed**: Closes #5794

**Tests performed**: (Does it build, tested in-game, etc)

By trickerer

### CODE REVIEW

**Comment by Ovahlord** [comment]
there has been several discussions about that feature, also that it was causing several crashes in multiple cases. I think this required loads of testing first

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -369,6 +377,19 @@ void Unit::Update(uint32 p_time)
         }
     }
 
+    for (uint8 i = BASE_ATTACK; i < RANGED_ATTACK; ++i)
+    {
+        if (_delayedTargetGuid[i])
+        {
+            if (_swingLanded[i] == true)
```
Comment (jackpoz) [inline src/server/game/Entities/Unit/Unit.cpp]: please don't == true, (_swingLanded[i]) is enough

**Comment by Aokromes** [comment]
@Ovahlord a crash was fixed over 3 years ago.

**Comment by Pixel-Pirate** [comment]
only for 335 branch ?

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -1999,23 +2057,27 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+            _swingDelayTimer[attType] = 500;
 
         if (GetTypeId() == TYPEID_PLAYER)
             TC_LOG_DEBUG("entities.unit", "AttackerStateUpdate: (Player) %u attacked %u (TypeId: %u) for %u dmg, absorbed %u, blocked %u, resisted %u.",
-                GetGUID().GetCounter(), victim->GetGUID().GetCounter(), victim->GetTypeId(), damageInfo.damage, damageInfo.absorb, damageInfo.blocked_amount, damageInfo.resist);
+                GetGUIDLow(), victim->GetGUIDLow(), victim->GetTypeId(), _damageInfo[attType].damage, _damageInfo[attType].absorb, _damageInfo[attType].blocked_amount, _damageInfo[attType].resist);
         else
             TC_LOG_DEBUG("entities.unit", "AttackerStateUpdate: (NPC)    %u attacked %u (TypeId: %u) for %u dmg, absorbed %u, blocked %u, resisted %u.",
-                GetGUID().GetCounter(), victim->GetGUID().GetCounter(), victim->GetTypeId(), damageInfo.damage, damageInfo.absorb, damageInfo.blocked_amount, damageInfo.resist);
+                GetGUIDLow(), victim->GetGUIDLow(), victim->GetTypeId(), _damageInfo[attType].damage, _damageInfo[attType].absorb, _damageInfo[attType].blocked_amount, _damageInfo[attType].resist);
```
Comment (Keader) [inline src/server/game/Entities/Unit/Unit.cpp]: GetGUIDLow() is not avaliable anymore i think
should be GetGUID()

Comment (Rochet2) [inline src/server/game/Entities/Unit/Unit.cpp]: Or maybe use `GetGUID().ToString().c_str()` and `%s` like most other log calls.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -1970,6 +1991,40 @@ void Unit::CalcHealAbsorb(Unit* victim, SpellInfo const* healSpell, uint32 &heal
...
+void Unit::ExecuteDelayedSwingHit(WeaponAttackType attType, bool extra)
+{
+    if (_swingLanded[attType])
+        return;
+
+    _swingLanded[attType] = true;
+
+    //if there is extra attack proc then target is already defined
+    if (!extra)
+        _damageInfo[attType].target = ObjectAccessor::GetUnit(*this, _delayedTargetGuid[attType]);
```
Comment (tkrokli) [inline src/server/game/Entities/Unit/Unit.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Unit/Unit.cpp:2003:70: fatal error: 
      reference to type 'const ObjectGuid' could not bind to an lvalue of type
      'uint64' (aka 'unsigned long')
  ...= ObjectAccessor::GetUnit(*this, _delayedTargetGuid[attType]);
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Globals/ObjectAccessor.h:77:69: note: 
      passing argument to parameter 'guid' here
    TC_GAME_API Unit* GetUnit(WorldObject const&, ObjectGuid const& guid);
                                                                    ^
```

Comment (Aokromes) [inline src/server/game/Entities/Unit/Unit.cpp]: known, pr needs big changes, because it's 2 years old code.

**Comment by SnapperRy** [comment]
I think this change (probably it'd better fit another PR, though) should affect spells, too. Right now, spells that kill a target don't have time to execute their animation, which I recall on retail was the case.

**Comment by Shauren** [comment]
Just checked retail, current behavior (damage and combat log event) happen when swing animation BEGINS, not hits. Scrolling white text on screen happens as expected when it hits both on TC and retail.

**Pull request closed without merging**